include/openPMD/backend/Attribute.hpp

```diff
@@ -61,4 +62,81 @@ using Attribute = auxiliary::Variant< Datatype,
                             std::vector< std::string >,
                             std::array< double, 7 >,
                             bool >;
-} // openPMD
+namespace detail
+{
+    template<bool wasFound, typename U, typename T_Head, typename ...T>
+    struct FindTypeImpl
+    {
+        static constexpr bool value = std::is_same<
+            U,
+            T_Head
+        >::value;
+
+        using type = typename std::conditional<
+            value,
+            T_Head,
+            typename FindTypeImpl<
+                wasFound || value,
+                U,
+                T...
+            >::type
+        >::type;
+    };
+
+    template<typename U, typename T, bool wasFound>
+    struct FindTypeImpl<wasFound, U, T>
+    {
+        static constexpr bool value = std::is_same<
+            U,
+            T
+        >::value;
+
+        // will throw below in case it is not T
+        using type = T;
+
+        static_assert(
+            wasFound || value,
+            "Type was not found!"
```

Comment:
Is this the compilation error one will get when doing something like
```cpp
struct NotAnAttributeDType
{ };

Series s = /* ...  */;
auto attr = NotAnAttributeDType();
s.setAttribute("nonsense", attr);
```
?

Reply:
Yes, e.g. on GCC 6.3 you get (I removed the c-string handling for demonstration):
```
openPMD-api/include/openPMD/backend/Attribute.hpp:97:9:
error: static assertion failed: Type was not found!
         static_assert(
         ^~~~~~~~~~~~~
In file included from openPMD-api/include/openPMD/Iteration.hpp:23:0,
                 from openPMD-api/include/openPMD/openPMD.hpp:27,
                 from openPMD-api/test/SerialIOTest.cpp:8:
openPMD-api/include/openPMD/backend/Attributable.hpp: In instantiation of ‘bool openPMD::Attributable::setAttribute(const string&, const T&) [with T = char [59]; std::__cxx11::string = std::__cxx11::basic_string<char>]’:
openPMD-api/test/SerialIOTest.cpp:1290:80:   required from here
openPMD-api/include/openPMD/backend/Attributable.hpp:234:32: warning: the compiler can assume that the address of ‘value’ will always evaluate to ‘true’ [-Waddress]
         it->second = Attribute(static_cast<FixedSize const&>(value));
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
openPMD-api/include/openPMD/backend/Attributable.hpp:240:66: warning: the compiler can assume that the address of ‘value’ will always evaluate to ‘true’ [-Waddress]
                                    std::make_pair(key, Attribute(static_cast<FixedSize const&>(value))));
```